airflow/models/taskinstance.py

```diff
@@ -2195,10 +2181,10 @@ def render_templates(self, context: Optional[Context] = None) -> "Operator":
         """
         if not context:
             context = self.get_template_context()
+        original_task = self.task
         rendered_task = self.task.render_template_fields(context)
-        if rendered_task is None:  # Compatibility -- custom renderer, assume unmapped.
-            return self.task
-        original_task, self.task = self.task, rendered_task
+        if rendered_task is not None:  # Mapped operator, assign unmapped task.
+            self.task = rendered_task
         return original_task
```

Comment:
… here. Previously it’s difficult to follow when `rendered_task` is None and when it’s not; now it’s simple—a mapped operator (MappedOperator subclases) performs unmapping and returns the unmapped task, while a non-mapped operator (BaseOperator subclasses) return None because no unmapping is needed.